### DISCUSSION TITLE
Use local downloaded bellsoft-liberica with binding tool

### DISCUSSION THREAD

**Discussion by ongiant** [paketo-buildpacks/feedback#246] [Java Team]
As [issue #323](https://github.com/paketo-buildpacks/bellsoft-liberica/issues/323) described, my internet connection is slow and unstable due to my network proxy maybe, so I  [dmikusa's advice](https://github.com/paketo-buildpacks/bellsoft-liberica/issues/323#issuecomment-1893786517). It succeed for `Syft` but it failed for `bellsoft-liberica`.
1.  I tried to create `dependency-mapping` for `bellsoft-liberica`, but it tried to download the `jre21 version` rather than `jre17`
<code><pre>
➜  binding_tool-x86_64-unknown-linux-gnu git:(master) ✗ ./bt dependency-mapping -b paketo-buildpacks/bellsoft-liberica@v10.4.4   
thread '<unnamed>' panicked at src/deps.rs:128:33:
Download of https://github.com/bell-sw/Liberica/releases/download/21.0.1+12/bellsoft-jdk21.0.1+12-linux-amd64.tar.gz failed with error https://github.com/bell-sw/Liberica/releases/download/21.0.1+12/bellsoft-jdk21.0.1+12-linux-amd64.tar.gz: Connection Failed: tls connection init failed: Resource temporarily unavailable (os error 11)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread '<unnamed>' panicked at src/deps.rs:125:45:
unable to get lock: PoisonError { .. }
thread '<unnamed>' panicked at src/deps.rs:125:45:
unable to get lock: PoisonError { .. }
thread '<unnamed>' panicked at src/deps.rs:125:45:
unable to get lock: PoisonError { .. }
thread '<unnamed>' panicked at src/deps.rs:125:45:
unable to get lock: PoisonError { .. }
Error: thread panic: unable to get lock: PoisonError { .. }
</pre></code>
2. `./gradlew bootBuildImage` command output logs screenshot:
![Screenshot_20240117_032537](https://github.com/paketo-buildpacks/feedback/assets/38517624/cd5593cf-3b57-42a7-b761-469b7ffe37dc)

**Question:** What should I do to handle this case? Alternatively, can the [binding tool](https://github.com/dmikusa/binding-tool?tab=readme-ov-file#binding-tool) create a `dependency mapping` utilizing the specified version (e.g. `BellSoft Liberica JRE 17.0.9`) of local `BellSoft Liberica` buildpacks?

**Comment by dmikusa** [comment]
Binding tool, if you feed it a buildpack, will download all of the dependencies for the buildpack. It doesn't presently allow you to filter them (you're welcome to open an issue for that).

The Bellsoft Liberica buildpack is pretty big and has a lot of dependencies, it could take a while to download them all if you have a slow Internet connection. On the plus side, you only need to do that once.

I don't know why you'd be getting that TLS error, maybe something with your proxy? Did you double-check your [proxy settings](https://github.com/dmikusa/binding-tool?tab=readme-ov-file#proxy-support)? Binding tool should honor them.

You could also do the downloads manually. Then you can only download what you want and you can use a tool that might be able to restart downloads, which is helpful on unstable Internet connections.

After you download the files manually, you can use step 2.) [here](https://github.com/dmikusa/binding-tool?tab=readme-ov-file#add-any-type-of-binding) to create the binding manually. The hash there is sha256, and `shasum -a 256` will generate that value. 

The trick is that you need to manage where it will expect the actual binaries to exist in the container. In the example there, it is using `file:///deps/bellsoft-jdk8u302+8-linux-amd64.tar.gz` as the path which would expect you to have one volume mount for your bindings, i.e. `/bindings`, and one volume mount at `/deps` with your binaries. When the binding tool downloads and generates the bindings automatically, it will put them both under `/bindings` that way you only need one volume mount.

**Reply by ongiant** [reply]
Great! It worked successfully after I correctly input the network proxy into the environment variable. Thank you! God bless you!

**Reply by dmikusa** [reply]
Excellent! Happy to help.
